### PR TITLE
Queryify `is_doc_hidden`

### PR DESCRIPTION
**Pull request by camelid**
It came up hot on some profiling of rustdoc I did, so hopefully turning
it into a query will help.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/util.rs [hunk 1]
```diff
@@ -1154,6 +1154,14 @@ pub fn normalize_opaque_types<'tcx>(
...
+pub fn is_doc_hidden(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+    tcx.get_attrs(def_id)
+        .iter()
+        .filter_map(|attr| if attr.has_name(sym::doc) { attr.meta_item_list() } else { None })
+        .any(|items| items.iter().any(|item| item.has_name(sym::hidden)))
+}
+
 pub fn provide(providers: &mut ty::query::Providers) {
-    *providers = ty::query::Providers { normalize_opaque_types, ..*providers }
+    *providers = ty::query::Providers { normalize_opaque_types, is_doc_hidden, ..*providers }
```
Comment (jyn514) [inline compiler/rustc_middle/src/ty/util.rs:1166]: this only provides it for the local crate I think, you should double check it doesn't crash on external crates

Comment (camelid) [reply compiler/rustc_middle/src/ty/util.rs:1166]: Huh, I didn't realize that, thanks!

Comment (cjgillot) [reply compiler/rustc_middle/src/ty/util.rs:1166]: The distinction between local and extern query providers only exists if the query is markes `separate_provide_extern`. This query is not marked as such, so this implementation will be called for both the local crate and extern crates.

**Comment by cjgillot** [comment]
The handling of attributes, implemented using a `iter().filter` seems generally inefficient. I filed https://github.com/rust-lang/rust/issues/94905 to have it refactored.
Meanwhile, this looks good, with a decent perf improvement. Great!
@bors r+

**Pull request merged**